File: link/sdks/dart/lib/src/cell_value.dart
import 'package:kalam_link/src/file_ref.dart';

/// Type-safe wrapper for a single cell value in a KalamDB query result row.
///
/// Mirrors `KalamCellValue` in Rust (`kalam-link/src/models/cell_value.rs`).
///
/// Every cell in a [QueryResult] row (and subscription / consume events) is
/// wrapped in a [KalamCellValue] so you can safely extract typed data:
///
/// ```dart
/// final name   = row['name']?.asString();     // String?
/// final age    = row['age']?.asInt();          // int?
/// final score  = row['score']?.asDouble();     // double?
/// final active = row['active']?.asBool();      // bool?
/// final ref    = row['avatar']?.asFile();      // KalamFileRef?
/// final url    = row['avatar']?.asFileUrl(
///   'http://localhost:18080', 'default', 'users',
/// );
/// ```
///
/// Use [KalamCellValue.from] to wrap a raw dynamic value returned by JSON
/// deserialization.
class KalamCellValue {
  /// The underlying raw value (may be `null`, `String`, `num`, `bool`,
  /// `Map<String, dynamic>`, or `List<dynamic>`).
  final dynamic _raw;

  /// Create a [KalamCellValue] wrapping [raw].
  ///
  /// Prefer the [KalamCellValue.from] factory for convenience.
  const KalamCellValue(this._raw);

  // ------------------------------------------------------------------ //
  //  Factory                                                            //
  // ------------------------------------------------------------------ //

  /// Wrap a raw dynamic value as a [KalamCellValue].
  ///
  /// Pass `null` for SQL NULL cells.
  static KalamCellValue from(dynamic raw) => KalamCellValue(raw);

  // ------------------------------------------------------------------ //
  //  Raw access                                                         //
  // ------------------------------------------------------------------ //

  /// Return the underlying raw JSON value.
  ///
  /// Use this when you need to pass the value to code that expects plain
  /// dynamic / JSON data (e.g. [KalamFileRef.tryParse]).
  dynamic toJson() => _raw;

  /// Return the underlying raw JSON value.
  ///
  /// Alias of [toJson] for teams that prefer `asX()` style accessors.
  dynamic asJson() => _raw;

  // ------------------------------------------------------------------ //
  //  Type guards                                                        //
  // ------------------------------------------------------------------ //

  /// `true` if this cell is SQL NULL.
  bool get isNull => _raw == null;

  /// `true` if the underlying value is a [String].
  bool get isString => _raw is String;

  /// `true` if the underlying value is a [num] (int or double).
  bool get isNumber => _raw is num;

  /// `true` if the underlying value is a [bool].
  bool get isBool => _raw is bool;

  /// `true` if the underlying value is a `Map<dynamic, dynamic>`.
  bool get isObject => _raw is Map;

  /// `true` if the underlying value is a [List].
  bool get isArray => _raw is List;

  // ------------------------------------------------------------------ //
  //  Typed accessors                                                    //
  // ------------------------------------------------------------------ //

  /// Return the value as a [String], or `null` for SQL NULL.
  ///
  /// Handles common Rust-side string envelopes like `{"Utf8": "..."}`.
  /// Numbers and booleans are converted to their string representation.
  String? asString() {
    if (isNull) return null;

    // Envelope: { "Utf8": "..." } or { "String": "..." }
    if (_raw is Map) {
      final map = _raw;
      if (map['Utf8'] is String) return map['Utf8'] as String;
      if (map['String'] is String) return map['String'] as String;
      return _raw.toString();
    }

    if (_raw is String) return _raw;
    if (_raw is num) return _raw.toString();
    if (_raw is bool) return _raw ? 'true' : 'false';
    return _raw.toString();
  }

  /// Return the value as an [int], or `null` for SQL NULL / non-integer.
  ///
  /// Truncates floating-point values. String-encoded integers are parsed.
  int? asInt() {
    if (isNull) return null;
    if (_raw is int) return _raw;
    if (_raw is double) return (_raw).truncate();
    if (_raw is bool) return _raw ? 1 : 0;
    if (_raw is String) {
      return int.tryParse(_raw) ?? double.tryParse(_raw)?.truncate();
    }
    return null;
  }

  /// Return the value as a [double], or `null` for SQL NULL / non-numeric.
  ///
  /// String-encoded floats are parsed.
  double? asDouble() {
    if (isNull) return null;
    if (_raw is double) return _raw;
    if (_raw is int) return (_raw).toDouble();
    if (_raw is bool) return _raw ? 1.0 : 0.0;
    if (_raw is String) return double.tryParse(_raw);
    return null;
  }

  /// Return the value as a [bool], or `null` for SQL NULL / non-boolean.
  ///
  /// Handles string-encoded booleans: `"true"`, `"false"`, `"1"`, `"0"`.
  bool? asBool() {
    if (isNull) return null;
    if (_raw is bool) return _raw;
    if (_raw is num) return _raw != 0;
    if (_raw is String) {
      final lc = (_raw).toLowerCase().trim();
      if (lc == 'true' || lc == '1') return true;
      if (lc == 'false' || lc == '0') return false;
    }
    return null;
  }

  /// Return the value as a [DateTime], or `null` for SQL NULL / unparseable.
  ///
  /// Handles:
  /// - Unix milliseconds as [int]
  /// - Unix microseconds as [int] (Postgres TIMESTAMPTZ wire format)
  /// - ISO 8601 strings (`"2024-01-01T00:00:00Z"`)
  /// - Numeric timestamp strings (`"1704067200000"` ms or `"1704067200000000"` µs)
  DateTime? asDate() {
    if (isNull) return null;
    if (_raw is int) {
      return _fromNumericTimestamp(_raw);
    }
    if (_raw is double) {
      return _fromNumericTimestamp((_raw).toInt());
    }
    if (_raw is String) {
      final raw = _raw;
      // Numeric timestamp string
      final n = int.tryParse(raw) ?? double.tryParse(raw)?.toInt();
      if (n != null) {
        return _fromNumericTimestamp(n);
      }
      return DateTime.tryParse(raw);
    }
    return null;
  }

  /// Converts a numeric timestamp to [DateTime].
  ///
  /// KalamDB server sends Postgres TIMESTAMPTZ values as microseconds since
  /// the Unix epoch (16-digit numbers for current dates). Plain milliseconds
  /// (13-digit numbers) are also accepted for backwards compatibility.
  ///
  /// Threshold: values > 9_999_999_999_999 are microseconds; anything smaller
  /// is treated as milliseconds. Year 2001 in ms ≈ 1_000_000_000_000 (13
  /// digits); year 2001 in µs ≈ 1_000_000_000_000_000 (16 digits), so there
  /// is no ambiguity for any date between 1970 and year 2286.
  static DateTime _fromNumericTimestamp(int n) {
    const msThreshold = 9999999999999;
    if (n.abs() > msThreshold) {
      return DateTime.fromMicrosecondsSinceEpoch(n, isUtc: true);
    }
    return DateTime.fromMillisecondsSinceEpoch(n, isUtc: true);
  }

  /// Return the value as a `Map<String, dynamic>`, or `null` if not an object.
  Map<String, dynamic>? asObject() {
    if (_raw is Map<String, dynamic>) return _raw;
    if (_raw is Map) {
      try {
        return Map<String, dynamic>.from(_raw);
      } catch (_) {
        return null;
      }
    }
    return null;
  }

  /// Return the value as a `List<dynamic>`, or `null` if not an array.
  List<dynamic>? asArray() {
    if (_raw is List) return _raw;
    return null;
  }

  // ------------------------------------------------------------------ //
  //  FILE column support                                                //
  // ------------------------------------------------------------------ //

  /// Parse a FILE column value and return a [KalamFileRef], or `null`.
  ///
  /// The FILE column stores a serialised JSON object matching `FileRefData`.
  /// This accessor deserialises it into a class instance with helper methods.
  ///
  /// ```dart
  /// final fileRef = row['avatar']?.asFile();
  /// if (fileRef != null) {
  ///   final url = fileRef.getDownloadUrl('http://localhost:18080', 'default', 'users');
  ///   print('${fileRef.name} (${fileRef.formatSize()})');
  /// }
  /// ```
  KalamFileRef? asFile() {
    if (isNull) return null;
    return KalamFileRef.tryParse(_raw);
  }

  /// Convenience: parse a FILE column and return the download URL directly.
  ///
  /// Returns `null` if the cell is null, not a valid file reference, or no
  /// URL can be built.
  ///
  /// - [baseUrl]   Server base URL (e.g. `"http://localhost:18080"`)
  /// - [namespace] Namespace of the table (e.g. `"default"`)
  /// - [table]     Table name (e.g. `"users"`)
  ///
  /// ```dart
  /// final url = row['avatar']?.asFileUrl(
  ///   'http://localhost:18080', 'default', 'users',
  /// );
  /// image.src = url ?? '/placeholder.png';
  /// ```
  String? asFileUrl(String baseUrl, String namespace, String table) {
    return asFile()?.getDownloadUrl(baseUrl, namespace, table);
  }

  // ------------------------------------------------------------------ //
  //  Serialisation / display                                            //
  // ------------------------------------------------------------------ //

  /// Human-readable string for display purposes.
  ///
  /// - SQL NULL → `"NULL"`
  /// - [String]  → value as-is
  /// - everything else → `_raw.toString()`
  @override
  String toString() {
    if (isNull) return 'NULL';
    if (_raw is String) return _raw;
    return _raw.toString();
  }

  @override
  bool operator ==(Object other) {
    if (other is KalamCellValue) return _raw == other._raw;
    return false;
  }

  @override
  int get hashCode => _raw.hashCode;
}

File: link/sdks/typescript/src/cell_value.ts
/**
 * KalamCellValue — Type-safe wrapper for individual cell values in KalamDB rows.
 *
 * ## Overview
 *
 * `KalamCellValue` wraps the raw JSON value returned by the server for a single
 * cell and provides typed accessor methods that mirror the Rust `KalamCellValue`
 * newtype (`kalam-link/src/models/cell_value.rs`).
 *
 * Instead of receiving `Record<string, unknown>` rows, consumers now receive
 * `RowData = Record<string, KalamCellValue>` rows and call safe accessors:
 *
 * ```typescript
 * const name   = row.name.asString();    // string | null
 * const age    = row.age.asInt();        // number | null
 * const active = row.active.asBool();   // boolean | null
 * const avatar = row.avatar.asFile();   // FileRef | null
 * const url    = row.avatar.asFileUrl('http://localhost:8080', 'default', 'users');
 * ```
 *
 * ## FILE columns
 *
 * `asFile()` parses the FILE column JSON and returns a `FileRef` instance, or
 * `null` if the cell is null or the value is not a valid file reference.
 * `asFileUrl()` is a convenience shorthand that builds the download URL directly.
 *
 * ## Utilities
 *
 * - `KalamCellValue.from(raw)` — wraps any raw JS value
 * - `wrapRowMap(raw)` — converts `Record<string, unknown>` → `RowData`
 *
 * @example
 * ```typescript
 * import { KalamCellValue, RowData } from 'kalam-link';
 *
 * // rows are already RowData[] from queryAll()
 * for (const row of rows) {
 *   console.log(row.id.asString(), row.score.asFloat());
 * }
 * ```
 *
 * @module
 */

import { FileRef } from './file_ref.js';
import { SeqId } from './seq_id.js';
import type { JsonValue } from './types.js';

/* ================================================================== */
/*  KalamCellValue class                                              */
/* ================================================================== */

/**
 * A type-safe wrapper for a single cell value in a KalamDB query result row.
 *
 * Mirrors `KalamCellValue` in Rust (`kalam-link/src/models/cell_value.rs`).
 */
export class KalamCellValue {
  /** @internal Raw JSON value as returned by the server */
  readonly #raw: JsonValue;

  /** @internal */
  private constructor(raw: JsonValue) {
    this.#raw = raw;
  }

  /* ------------------------------------------------------------------ */
  /*  Factory                                                            */
  /* ------------------------------------------------------------------ */

  /**
   * Wrap a raw JS value (from JSON deserialization) as a `KalamCellValue`.
   *
   * Pass `null` / `undefined` for SQL NULL cells.
   */
  static from(raw: unknown): KalamCellValue {
    // Normalise undefined → null (treat as SQL NULL)
    return new KalamCellValue(raw === undefined ? null : (raw as JsonValue));
  }

  /* ------------------------------------------------------------------ */
  /*  Raw access (mirrors Rust `.inner()`)                              */
  /* ------------------------------------------------------------------ */

  /**
   * Return the underlying raw JSON value.
   *
   * Use this when you need to pass the value to code that expects plain JSON
   * (e.g., existing WASM helpers, `FileRef.from(cell.toJson())`).
   */
  toJson(): JsonValue {
    return this.#raw;
  }

  /**
   * Return the underlying raw JSON value.
   *
   * Alias of `toJson()` for codebases that prefer `asX()` naming.
   */
  asJson(): JsonValue {
    return this.#raw;
  }

  /* ------------------------------------------------------------------ */
  /*  Type guards                                                       */
  /* ------------------------------------------------------------------ */

  /** Returns `true` if this cell is SQL NULL. */
  isNull(): boolean {
    return this.#raw === null || this.#raw === undefined;
  }

  /** Returns `true` if the underlying value is a JSON string. */
  isString(): boolean {
    return typeof this.#raw === 'string';
  }

  /** Returns `true` if the underlying value is a JSON number. */
  isNumber(): boolean {
    return typeof this.#raw === 'number';
  }

  /** Returns `true` if the underlying value is a JSON boolean. */
  isBool(): boolean {
    return typeof this.#raw === 'boolean';
  }

  /** Returns `true` if the underlying value is a JSON object (not array, not null). */
  isObject(): boolean {
    return (
      this.#raw !== null &&
      typeof this.#raw === 'object' &&
      !Array.isArray(this.#raw)
    );
  }

  /** Returns `true` if the underlying value is a JSON array. */
  isArray(): boolean {
    return Array.isArray(this.#raw);
  }

  /* ------------------------------------------------------------------ */
  /*  Typed accessors                                                   */
  /* ------------------------------------------------------------------ */

  /**
   * Return the value as a string, or `null` for SQL NULL.
   *
   * Handles the `{ "Utf8": "..." }` envelope that some Rust drivers emit and
   * converts numbers / booleans to their string representations.
   */
  asString(): string | null {
    if (this.isNull()) return null;

    // Envelope: { "Utf8": "..." }
    if (this.isObject()) {
      const obj = this.#raw as Record<string, JsonValue>;
      if ('Utf8' in obj && typeof obj['Utf8'] === 'string') return obj['Utf8'];
      if ('String' in obj && typeof obj['String'] === 'string') return obj['String'];
    }

    if (typeof this.#raw === 'string') return this.#raw;
    if (typeof this.#raw === 'number') return String(this.#raw);
    if (typeof this.#raw === 'boolean') return this.#raw ? 'true' : 'false';

    return JSON.stringify(this.#raw);
  }

  /**
   * Return the value as an integer (`number`), or `null` for SQL NULL / non-numeric.
   *
   * Truncates floating-point values (same as `as i64` in Rust).
   * String-encoded integers (e.g. `"42"`) are parsed.
   */
  asInt(): number | null {
    if (this.isNull()) return null;
    if (typeof this.#raw === 'number') return Math.trunc(this.#raw);
    if (typeof this.#raw === 'string') {
      const n = Number(this.#raw);
      return Number.isFinite(n) ? Math.trunc(n) : null;
    }
    if (typeof this.#raw === 'boolean') return this.#raw ? 1 : 0;
    return null;
  }

  /**
   * Return the value as a native `bigint`, or `null` for SQL NULL / non-numeric.
   *
   * Useful for `Int64` / `UInt64` columns where the value may exceed
   * `Number.MAX_SAFE_INTEGER`.  String-encoded integers are parsed.
   */
  asBigInt(): bigint | null {
    if (this.isNull()) return null;
    try {
      if (typeof this.#raw === 'bigint') return this.#raw;
      if (typeof this.#raw === 'number') return BigInt(Math.trunc(this.#raw));
      if (typeof this.#raw === 'string') return BigInt(this.#raw.trim());
    } catch {
      // non-integer string
    }
    return null;
  }

  /**
   * Return the value as a `SeqId`, or `null` for SQL NULL / non-numeric.
   *
   * Use this for `_seq` columns or any Snowflake-based sequence ID.
   *
   * @example
   * ```typescript
   * const seq = row._seq.asSeqId();
   * if (seq) {
   *   console.log(seq.timestampMillis()); // when the row was written
   *   console.log(seq.workerId());        // which worker generated it
   * }
   * ```
   */
  asSeqId(): SeqId | null {
    if (this.isNull()) return null;
    try {
      if (typeof this.#raw === 'number') return SeqId.from(this.#raw);
      if (typeof this.#raw === 'bigint') return SeqId.from(this.#raw);
      if (typeof this.#raw === 'string') return SeqId.from(this.#raw.trim());
    } catch {
      // not parseable as SeqId
    }
    return null;
  }

  /**
   * Return the value as a floating-point number, or `null` for SQL NULL / non-numeric.
   *
   * String-encoded floats (e.g. `"3.14"`) are parsed.
   */
  asFloat(): number | null {
    if (this.isNull()) return null;
    if (typeof this.#raw === 'number') return this.#raw;
    if (typeof this.#raw === 'string') {
      const n = parseFloat(this.#raw);
      return Number.isFinite(n) ? n : null;
    }
    if (typeof this.#raw === 'boolean') return this.#raw ? 1.0 : 0.0;
    return null;
  }

  /**
   * Return the value as a boolean, or `null` for SQL NULL / non-boolean.
   *
   * Handles string-encoded booleans (`"true"` / `"false"` / `"1"` / `"0"`).
   */
  asBool(): boolean | null {
    if (this.isNull()) return null;
    if (typeof this.#raw === 'boolean') return this.#raw;
    if (typeof this.#raw === 'number') return this.#raw !== 0;
    if (typeof this.#raw === 'string') {
      const lc = this.#raw.toLowerCase().trim();
      if (lc === 'true' || lc === '1') return true;
      if (lc === 'false' || lc === '0') return false;
    }
    return null;
  }

  /**
   * Return the value as a `Date`, or `null` for SQL NULL / unparseable values.
   *
   * Handles:
   * - Unix milliseconds (number)
   * - ISO 8601 strings (`"2024-01-01T00:00:00Z"`)
   * - Numeric timestamp strings (`"1704067200000"`)
   */
  asDate(): Date | null {
    if (this.isNull()) return null;
    if (typeof this.#raw === 'number') {
      const d = new Date(this.#raw);
      return isNaN(d.getTime()) ? null : d;
    }
    if (typeof this.#raw === 'string') {
      // Try numeric timestamp first
      const n = Number(this.#raw);
      if (Number.isFinite(n)) {
        const d = new Date(n);
        return isNaN(d.getTime()) ? null : d;
      }
      const d = new Date(this.#raw);
      return isNaN(d.getTime()) ? null : d;
    }
    return null;
  }

  /**
   * Return the raw value as a plain JSON object, or `null`.
   */
  asObject(): Record<string, JsonValue> | null {
    if (this.isObject()) return this.#raw as Record<string, JsonValue>;
    return null;
  }

  /**
   * Return the raw value as a JSON array, or `null`.
   */
  asArray(): JsonValue[] | null {
    if (this.isArray()) return this.#raw as JsonValue[];
    return null;
  }

  /* ------------------------------------------------------------------ */
  /*  FILE column support                                               */
  /* ------------------------------------------------------------------ */

  /**
   * Parse a FILE column value and return a `FileRef` instance, or `null`.
   *
   * The FILE column stores a serialised JSON object matching `FileRefData`.
   * This accessor deserialises it into a class instance with helper methods.
   *
   * @example
   * ```typescript
   * const fileRef = row.avatar.asFile();
   * if (fileRef) {
   *   const url = fileRef.getDownloadUrl('http://localhost:8080', 'default', 'users');
   *   console.log(fileRef.name, fileRef.formatSize(), fileRef.isImage());
   * }
   * ```
   */
  asFile(): FileRef | null {
    if (this.isNull()) return null;
    return FileRef.from(this.#raw);
  }

  /**
   * Convenience: parse a FILE column and return the download URL directly.
   *
   * Returns `null` if the cell is null, not a valid file reference, or the
   * file ref does not produce a URL.
   *
   * @param baseUrl   - Server base URL (e.g. `"http://localhost:8080"`)
   * @param namespace - Namespace of the table (e.g. `"default"`)
   * @param table     - Table name (e.g. `"users"`)
   *
   * @example
   * ```typescript
   * const url = row.avatar.asFileUrl('http://localhost:8080', 'default', 'users');
   * img.src = url ?? '/placeholder.png';
   * ```
   */
  asFileUrl(baseUrl: string, namespace: string, table: string): string | null {
    const ref = this.asFile();
    if (!ref) return null;
    return ref.getDownloadUrl(baseUrl, namespace, table);
  }

  /* ------------------------------------------------------------------ */
  /*  Serialisation / display                                           */
  /* ------------------------------------------------------------------ */

  /**
   * Human-readable string for display purposes.
   *
   * - SQL NULL → `"NULL"`
   * - strings   → value as-is
   * - objects / arrays → compact JSON
   * - everything else → `String(value)`
   */
  toString(): string {
    if (this.isNull()) return 'NULL';
    if (typeof this.#raw === 'string') return this.#raw;
    if (typeof this.#raw === 'object') return JSON.stringify(this.#raw);
    return String(this.#raw);
  }
}

/* ================================================================== */
/*  RowData type                                                      */
/* ================================================================== */

/**
 * A single query result row with all values wrapped as `KalamCellValue`.
 *
 * Replaces `Record<string, unknown>` for strongly-typed row access.
 */
export type RowData = Record<string, KalamCellValue>;

/* ================================================================== */
/*  Utility helpers                                                   */
/* ================================================================== */

/**
 * Wrap a single `Record<string, unknown>` object row as `RowData`,
 * converting each value to a `KalamCellValue`.
 *
 * @example
 * ```typescript
 * const typedRow = wrapRowMap(rawRow);
 * console.log(typedRow.name.asString());
 * ```
 */
export function wrapRowMap(raw: Record<string, unknown>): RowData {
  const result: RowData = {};
  for (const key of Object.keys(raw)) {
    result[key] = KalamCellValue.from(raw[key]);
  }
  return result;
}

